elements/lisk-chain/src/block_header.ts

```diff
@@ -0,0 +1,164 @@
+	readonly generatorAddress: Buffer;
+	readonly previousBlockID: Buffer;
+	readonly timestamp: number;
+	readonly stateRoot: Buffer;
+	readonly transactionRoot: Buffer;
+	readonly version: number;
+	readonly assets: ReadonlyArray<{ moduleID: number; data: Buffer }>;
+	signature?: Buffer;
+	id?: Buffer;
+}
+
+export interface BlockHeaderJSON {
+	readonly generatorAddress: string;
+	readonly previousBlockID: string;
+	readonly timestamp: number;
+	readonly stateRoot: string;
+	readonly transactionRoot: string;
+	readonly version: number;
+	readonly assets: ReadonlyArray<{ moduleID: number; data: string }>;
+	readonly signature: string;
+	readonly id: string;
+}
+
+export class BlockHeader {
+	public readonly generatorAddress: Buffer;
+	public readonly version: number;
+	public readonly previousBlockID: Buffer;
+	public readonly timestamp: number;
+	public readonly stateRoot: Buffer;
+	public readonly transactionRoot: Buffer;
+
+	private readonly _assets: { moduleID: number; data: Buffer }[];
+	private _signature?: Buffer;
+	private _id?: Buffer;
+
+	public constructor({
+		version,
+		generatorAddress,
+		previousBlockID,
+		timestamp,
+		stateRoot,
+		transactionRoot,
+		signature,
+		id,
+		assets,
+	}: BlockHeaderAttrs) {
+		this.version = version;
+		this.generatorAddress = generatorAddress;
+		this.previousBlockID = previousBlockID;
+		this.timestamp = timestamp;
+		this.stateRoot = stateRoot;
+		this.transactionRoot = transactionRoot;
+		this._assets = [...assets];
+
+		this._signature = signature;
+		this._id = id;
+	}
+
+	public static fromBytes(value: Buffer): BlockHeader {
+		return new BlockHeader(codec.decode<BlockHeaderAttrs>(blockHeaderSchema, value));
+	}
+
+	public getBytes(): Buffer {
+		return codec.encode(blockHeaderSchema, this._getBlockHeaderProps());
+	}
+
+	public toJSON(): BlockHeaderJSON {
+		return codec.toJSON(blockHeaderSchema, this._getBlockHeaderProps());
+	}
+
+	public validate(): void {
+		const headerWithoutAsset = {
+			...this._getBlockHeaderProps(),
+			asset: Buffer.alloc(0),
+		};
+		// Validate block header
+		const errors = validator.validate(blockHeaderSchema, headerWithoutAsset);
+		if (errors.length) {
+			throw new LiskValidationError(errors);
+		}
+
+		if (this.previousBlockID.length === 0) {
+			throw new Error('Previous block id must not be empty');
+		}
+	}
+
+	public getSigningBytes(): Buffer {
+		const blockHeaderBytes = codec.encode(blockHeaderSchema, ({
+			...this,
+			signature: Buffer.alloc(0),
+		} as unknown) as Record<string, unknown>);
+
+		return blockHeaderBytes;
+	}
+
+	public sign(key: Buffer, networkIdentifier: Buffer) {
+		return signDataWithPrivateKey(TAG_BLOCK_HEADER, networkIdentifier, this.getSigningBytes(), key);
+	}
+
+	public getAsset(moduleID: number): Buffer | undefined {
+		return this._assets.find(a => a.moduleID === moduleID)?.data;
+	}
+	public setAsset(moduleID: number, value: Buffer): void {
+		const asset = this.getAsset(moduleID);
+		if (asset) {
```

Comment:
should we overwrite?

Reply:
Approach 1: Different modules can set asset on different logic, so we can overwrite asset. 
Approach 2: Let user handle logic on their side and set the asset once for a module. 

I believe Approach 2 is safe.

Reply:
I was thinking a case where a module want to set in `initBlock` and update in `sealBlock`, but i guess it's not necessary for now